### PR TITLE
[PVR] Alternative pvr announce

### PR DESCRIPTION
**Pull request by janbar**
Alternative to handle announcements to pvr addon. See PR #9654.

@ksooo @Jalle19 @FernetMenta

### CODE REVIEW

**Comment by janbar** [comment]
It works with addon pvr.mythtv, and it will work too with others after adaptation.

File: xbmc/pvr/PVRManager.cpp [hunk 1]
```diff
@@ -113,23 +113,38 @@ CPVRManager::~CPVRManager(void)
...
 void CPVRManager::Announce(AnnouncementFlag flag, const char *sender, const char *message, const CVariant &data)
 {
-  if (!IsStarted() || (flag & (System)) == 0)
+  if (!IsStarted())
     return;
 
-  if (strcmp(message, "OnWake") == 0)
+  if ((flag & (ANNOUNCEMENT::System)))
+  {
+    if (strcmp(message, "OnSleep") == 0)
```
Comment (FernetMenta) [inline xbmc/pvr/PVRManager.cpp]: this is not reliable. there is no guarantee that async announce messages get delivered before system gets down. you may get this message on wakeup.
that why I said the the IAnnounce interface was abused for those purposes. System behaviour should not depend on messages coming through announce.

Better do something like g_PVRManager.OnSleep/OnWake/etc

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h [hunk 2]
```diff
@@ -77,10 +77,10 @@ struct DemuxPacket;
 #define PVR_STREAM_MAX_STREAMS 20
 
 /* current PVR API version */
-#define XBMC_PVR_API_VERSION "5.1.1"
+#define XBMC_PVR_API_VERSION "6.0.0"
```
Comment (FernetMenta) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h]: no sure if we should bump major

Comment (MartijnKaijser) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h]: I'd say 5.2.0 should be enough as it's not a total breaking change

Comment (ksooo) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h]: I suggested this, but I think I was wrong. Minor bump should be the right thing to do for new api functions, that are actually called by Kodi, right?

File: xbmc/pvr/PVRManager.cpp (hunk 1 above)
Comment (ksooo) [inline xbmc/pvr/PVRManager.cpp]: @janbar Yeah, unfortunately, changing it the way you did would have exactly the problem we are trying to solve with removal if the old api function.

**Comment by janbar** [comment]
When I have a couple of minutes , I will update it regarding your comments.

File: xbmc/pvr/PVRManager.cpp (hunk 1 above)
Comment (janbar) [inline xbmc/pvr/PVRManager.cpp]: Not exactly. Here addon functions are called by the owner.

File: xbmc/pvr/PVRManager.cpp [hunk 3]
```diff
@@ -113,23 +113,34 @@ CPVRManager::~CPVRManager(void)
...
-    /* continue last watched channel */
-    ContinueLastChannel();
-
-    /* trigger PVR data updates */
-    TriggerChannelGroupsUpdate();
-    TriggerChannelsUpdate();
-    TriggerRecordingsUpdate();
-    TriggerEpgsCreate();
-    TriggerTimersUpdate();
+    if (strcmp(message, "OnWake") == 0)
```
Comment (Jalle19) [inline xbmc/pvr/PVRManager.cpp]: Shouldn't this be moved to the newly introduced `OnWake()` method?

File: xbmc/pvr/addons/PVRClients.cpp [hunk 4]
```diff
@@ -1615,3 +1615,43 @@ void CPVRClients::ConnectionStateChange(int clientId, std::string &strConnection
...
   }
 }
+
+void CPVRClients::OnSystemSleep()
+{
+  PVR_CLIENTMAP clients;
+  GetCreatedClients(clients);
+
+  /* propagate announcement to each client */
+  for (PVR_CLIENTMAP_CITR itrClients = clients.begin(); itrClients != clients.end(); itrClients++)
```
Comment (Jalle19) [inline xbmc/pvr/addons/PVRClients.cpp]: `for (auto &client : clients)`

File: xbmc/pvr/PVRManager.cpp [hunk 5]
```diff
@@ -466,6 +477,22 @@ bool CPVRManager::SetWakeupCommand(void)
   return false;
 }
 
+void CPVRManager::OnSleep()
+{
+  if (!IsStarted())
```
Comment (FernetMenta) [inline xbmc/pvr/PVRManager.cpp]: IsStarted is wrong. PVRmanager does not own the addons anymore

Comment (janbar) [inline xbmc/pvr/PVRManager.cpp]: Here the test checks is the PVRManager has fully started, not addon. What do you think to check instead ?

File: xbmc/pvr/PVRManager.cpp (hunk 3 above)
Comment (janbar) [inline xbmc/pvr/PVRManager.cpp]: Depends. It designed like this for some reason I don't know.

File: xbmc/pvr/PVRManager.cpp (hunk 5 above)
Comment (FernetMenta) [inline xbmc/pvr/PVRManager.cpp]: nothing to check here. PVRManager's state is of no interest.

File: xbmc/pvr/PVRManager.cpp (hunk 3 above)
Comment (FernetMenta) [inline xbmc/pvr/PVRManager.cpp]: I would leave this here. It is something that can be executed async.

**Comment by janbar** [comment]
Still long story with pvr addons. I will prepare their PR in the wake.

**Comment by janbar** [comment]
So now all is ready to be triggered.

**Comment by FernetMenta** [comment]
Thank you very much!
jenkins build this please

**Comment by ksooo** [comment]
Thanks @janbar for this and the monkey work with the add-ons.

**Comment by stefansaraev** [comment]
it seems you forgot to update addons/xbmc.pvr/addon.xml

**Pull request merged**